### PR TITLE
feat(OverlayTrigger): allow renderProp pattern

### PR DESCRIPTION
**Pull request by jquense**
Adds 2 features. First, OverlayTrigger render prop for more nuanced positioning situations (added docs example). Second, adds a global shared delay for tooltips in toolbars.

### CODE REVIEW

File: src/useDelayToggleCallback.tsx [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+let showHandle: NodeJS.Timeout;
+let leaveHandle: NodeJS.Timeout;
+const hideCallbacks = [] as Array<(nextValue: boolean) => void>;
+
+function drainHideCallbacks() {
+  hideCallbacks.forEach((h) => h(false));
+  hideCallbacks.length = 0;
+}
+
+function useGlobalTimeout(
```
Comment (jquense) [inline src/useDelayToggleCallback.tsx:23]: this is extremely inelegant, open to simplifications.

**Comment by kosmiq** [comment]
Super nice features!

I'll leave a proper review soon-ish. :)

File: src/useDelayToggleCallback.tsx [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+type ValueRef = {
+  entered?: boolean;
+  pending?: boolean;
+  hideCallbacks: Array<() => void>;
+};
+
+let entered: boolean;
+let showHandle: NodeJS.Timeout;
+let leaveHandle: NodeJS.Timeout;
```
Comment (taion) [inline src/useDelayToggleCallback.tsx:15]: **question:** are these going to cause problems w/people not using node types?

File: src/useDelayToggleCallback.tsx [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+function useGlobalTimeout(
+  delay: Delay,
+  callback: (nextValue: boolean) => void,
+) {
+  const [showDelay, leaveDelay] = delay;
+  return useCallback(
+    (nextValue: boolean) => {
+      clearTimeout(showHandle);
+
+      if (nextValue === true) {
```
Comment (taion) [inline src/useDelayToggleCallback.tsx:32]: **thought:** seems pointless if we know `nextValue` is a bool?

```suggestion
      if (nextValue) {
```

File: src/useDelayToggleCallback.tsx [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+function drainHideCallbacks() {
+  hideCallbacks.forEach((h) => h(false));
+  hideCallbacks.length = 0;
+}
+
+function useGlobalTimeout(
+  delay: Delay,
+  callback: (nextValue: boolean) => void,
+) {
+  const [showDelay, leaveDelay] = delay;
```
Comment (taion) [inline src/useDelayToggleCallback.tsx:27]: **nitpick:** could destructure

```suggestion
  [showDelay, leaveDelay]: Delay,
  callback: (nextValue: boolean) => void,
) {
```

**Comment by taion** [comment]
btw is the global thing present upstream?

**Comment by jquense** [comment]
> if you're in the "show" state, no delay in switching from one to the other

That's the core of it yes, Once a tooltip is shown you want the next one to show immediately until an normal delayed exit occurs. 

I threw this together very quickly b/c it was annoying me and i was hoping i could avoid a more involved API. 

something like: [tippy](https://atomiks.github.io/tippyjs/#singleton) or the reach-ui tooltip

I think it's ok UX to not have more local or nuanced scoping, like a provider, but I agree that as is it's ugly and bad. Like you probably also want to skip the animation for already entered tips but a good way to do that didn't seem obvious. Maybe what you actually want to do is share a single tooltip component and change the content and reference, but that seems hard...

**Comment by jquense** [comment]
Ok I split out the global timer stuff for now to not hold this up

**Comment by syadav214** [comment]
UI is breaking for me on tooltip.

image

**Pull request merged**